src/Processors/Transforms/Streaming/WatermarkTransformWithSubstream.cpp

```diff
@@ -127,36 +127,39 @@ void WatermarkTransformWithSubstream::work()
         checkpoint(process_chunk.getCheckpointContext());
         output_chunks.emplace_back(std::move(process_chunk));
     }
-    else if (avoid_watermark)
-    {
-        output_chunks.emplace_back(std::move(process_chunk));
-    }
     else if (process_chunk.hasRows())
     {
         assert(process_chunk.hasChunkContext());
 
         auto & watermark = getOrCreateSubstreamWatermark(process_chunk.getSubstreamID());
 
-        watermark.process(process_chunk);
-        assert(process_chunk);
+        if (!avoid_watermark)
+            watermark.process(process_chunk);
 
+        assert(process_chunk);
         output_chunks.emplace_back(std::move(process_chunk));
     }
     else
     {
         /// FIXME, we shall establish timer only when necessary instead of blindly generating empty heartbeat chunk
         bool propagated_heartbeat = false;
-        output_chunks.reserve(substream_watermarks.size());
-        for (auto & [id, watermark] : substream_watermarks)
-        {
-            auto chunk = process_chunk.clone();
-            watermark->process(chunk);
 
-            if (chunk.hasChunkContext())
+        /// It's possible to generate periodic or timeout watermark for each substream by empty chunk
+        /// FIXME: This is a very ugly and inefficient implementation and needs to revisit.
+        if (!avoid_watermark && watermark_template->requiresPeriodicOrTimeout())
```

Comment:
For substream, what does it mean for periodic / timeout emit ? Applied for each substream ?

Reply:
So far, for periodic/timeout emit, we always apply the special watermark to each substream.
Of course, it doesn't make sense for periodic emit, we can optimize it in the future